libia2/include/ia2.h

```diff
@@ -264,13 +254,86 @@ static int insecure_pkey_mprotect(void *ptr, size_t len, int prot, int pkey) {
       *n_to_alloc = 0;                                                         \
     }                                                                          \
   }                                                                            \
-  char *ia2_stackptrs[n + 1] IA2_SHARED_DATA;                                  \
-  __attribute__((constructor)) static void init_stacks() {                     \
-    ensure_pkeys_allocated(&ia2_n_pkeys_to_alloc);                             \
+                                                                               \
+  /* one page of stack pointers for each compartment */                        \
+  char *ia2_stackptrs[PAGE_SIZE * (n + 1)] __attribute__((aligned(PAGE_SIZE))) \
+  IA2_SHARED_DATA;                                                             \
+                                                                               \
+  static void protect_stackptrs(void) {                                        \
+    for (int i = 0; i < n + 1; i++) {                                          \
+      char *stackptrs_page = (char *)ia2_stackptrs + i * PAGE_SIZE;            \
+      int res =                                                                \
+          pkey_mprotect(stackptrs_page, PAGE_SIZE, PROT_READ | PROT_WRITE, i); \
+      if (res == -1) {                                                         \
+        printf("Failed to mprotect stack pointers for compartment %d (%s)\n",  \
+               i, errno_s);                                                    \
+        exit(1);                                                               \
+      }                                                                        \
+    }                                                                          \
+  }                                                                            \
+                                                                               \
+  /* Allocate and protect the stack for this thread's \p i'th compartment. */  \
+  /* Stores the stack in the appropriate slot of ia2_stackptrs. */             \
+                                                                               \
+  /* I'm worried about interference from other threads if any sensitive */     \
+  /* locals of in this function get spilled to the stack. We probably need */  \
+  /* to write this in assembly. */                                             \
```

Comment:
I agree with this comment. Also if `old_pkru` can be any compartment this function might trigger an mpk violation after the wrpkru if it access the stack. We could avoid that by switching to a mixed pkru, but I think it's better to avoid that when possible.

Reply:
Does that solve the problem though? Are there any ways that registers can be spilled underneath us in process memory?

Reply:
The problem of spilling onto the stack or are you thinking about things like thread-locals? I don't see what the issue is if we don't write to the stack in this asm function

Reply:
As far as I can tell that will solve the problem. We generally rely on registers being safe from outside interference in IA2, e.g. in stack switching and register scrubbing.